app/controllers/mixins/ems_common_angular.rb

```diff
@@ -266,6 +266,20 @@ def ems_form_fields
                         :ssh_keypair_auth_status       => ssh_keypair_auth_status.nil? ? true : ssh_keypair_auth_status
       } if controller_name == "ems_infra"
 
+      render :json => {
+        :name                => @ems.name,
+        :emstype             => @ems.emstype,
+        :zone                => zone,
+        :provider_id         => @ems.provider_id ? @ems.provider_id : "",
+        :hostname            => @ems.hostname,
+        :default_hostname    => @ems.connection_configurations.default.endpoint.hostname,
+        :default_api_port    => @ems.connection_configurations.default.endpoint.port,
+        :provider_region     => @ems.provider_region,
+        :default_userid      => @ems.authentication_userid ? @ems.authentication_userid : "",
+        :ems_controller      => controller_name,
+        :default_auth_status => default_auth_status,
+      } if controller_name == "ems_physical_infra"
```

Comment:
> The need to add new code that follows a bad pattern should be the trigger for refactoring :-(

@martinpovolny Could not agree more.
However, it goes without saying that the refactoring effort should not be part of this (already gigantic) PR. That said, what I can offer here is this commit that I just pushed:
1dd3e439ea4813b8c9a550df43fcee04480c7053 
It renders the `ems_physical_infra`-specific json using it's own controller method, making sure it does not add to the mess in the mixin.
Does that work?